pkg/executables/cmk.go

```diff
@@ -0,0 +1,363 @@
+		if err != nil {
+			return result, fmt.Errorf("error getting resource info filtering by id %s: %v", parameterValue, err)
+		}
+	}
+	return result, nil
+}
+
+func (c *Cmk) exec(ctx context.Context, args ...string) (stdout bytes.Buffer, err error) {
+	c.setupExecConfig()
+	envMap, err := c.getEnvMap()
+	if err != nil {
+		return bytes.Buffer{}, fmt.Errorf("failed get environment map: %v", err)
+	}
+	err = c.buildCmkConfigFile(envMap)
+	if err != nil {
+		return bytes.Buffer{}, fmt.Errorf("failed cmk validations: %v", err)
+	}
+	argsWithConfigFile := append([]string{"-c", c.execConfig.ConfigFile}, args...)
+
+	return c.executable.Execute(ctx, argsWithConfigFile...)
+}
+
+// TODO: Add support for passing in domain from Deployment Config Spec
+func (c *Cmk) buildCmkConfigFile(envMap map[string]string) (err error) {
+	decodedString, err := b64.StdEncoding.DecodeString(envMap[cloudStackb64EncodedSecretKey])
+	if err != nil {
+		return fmt.Errorf("failed to decode value for %s with base64: %v", cloudStackb64EncodedSecretKey, err)
+	}
+	cfg, err := ini.Load(decodedString)
+	if err != nil {
+		return fmt.Errorf("failed to extract values from %s with ini: %v", cloudStackb64EncodedSecretKey, err)
+	}
+	section, err := cfg.GetSection("Global")
+	if err != nil {
+		return fmt.Errorf("failed to extract section 'Global' from %s: %v", cloudStackb64EncodedSecretKey, err)
+	}
+	apiKey, err := section.GetKey("api-key")
+	if err != nil {
+		return fmt.Errorf("failed to extract value of 'api-key' from %s: %v", cloudStackb64EncodedSecretKey, err)
+	}
+	secretKey, err := section.GetKey("secret-key")
+	if err != nil {
+		return fmt.Errorf("failed to extract value of 'secret-key' from %s: %v", cloudStackb64EncodedSecretKey, err)
+	}
+	apiUrl, err := section.GetKey("api-url")
+	if err != nil {
+		return fmt.Errorf("failed to extract value of 'api-url' from %s: %v", cloudStackb64EncodedSecretKey, err)
+	}
+	cmkInsecure, err := strconv.ParseBool(envMap[cloudmonkeyInsecureKey])
+	if err != nil {
+		return fmt.Errorf("failed to parse boolean value from %s: %v", cloudmonkeyInsecureKey, err)
+	}
+	cmkVerifyCert := strconv.FormatBool(!cmkInsecure)
+	t := templater.New(c.writer)
+	data := map[string]string{
+		"CloudStackApiKey":        apiKey.Value(),
+		"CloudStackSecretKey":     secretKey.Value(),
+		"CloudStackManagementUrl": apiUrl.Value(),
+		"CloudMonkeyVerifyCert":   cmkVerifyCert,
+	}
+	writtenFileName, err := t.WriteToFile(cmkConfigTemplate, data, cmkConfigFileName)
+	if err != nil {
+		return fmt.Errorf("error creating file for cmk config: %v", err)
+	}
+	c.execConfig.ConfigFile, err = filepath.Abs(writtenFileName)
+	if err != nil {
+		return fmt.Errorf("failed to generate absolute filepath for generated config file at %s", writtenFileName)
+	}
+
+	return nil
+}
+
+func (c *Cmk) getEnvMap() (map[string]string, error) {
+	envMap := make(map[string]string)
+	for _, key := range requiredEnvsCmk {
+		if env, ok := os.LookupEnv(key); ok && len(env) > 0 {
```

Comment:
Could we avoid passing info in env vars? I know we have used this pattern in the past but it has proven to be a bit inflexible and cumbersome to expand

I would suggest passing all the necessary info as arguments to the `Cmk` constructor
If there are too many to be separate args, it's totally fine to create a `CmkConf` struct and pass it as input to the constructor

Reply:
To be clear, passing in data to the eks-a executable should still be done with environment variables, correct? You're just suggesting passing data through to Cmk specifically not be done with environment variables?

Reply:
Yeah, totally agree, we should pass the "secret" data to the eks-a cli through env vars. But that doesn't mean we need to read them from `Cmk`. We can just read them at a higher level (eg. the `cmd` package) and pass them down to the `Cmk` constructor.

My main reasons to suggest this:
* Making testing easier (instead of setting env vars in the test, we just pass strings to the constructor). This facilitates parallel testing.
* Portability: we will be able to use this from a controller, where the secrets won't come from env vars (most probably from something like kubernetes secrets)
* Flexibility: it will allow us to have multiple `Cmk` instances, each with a different set of credentials (this could be useful when having to contact multiple "datacenters")

I know none of the things I just mentioned are necessarily requirements for this current usecase. But we learnt with `govc` that using env vars at the executable level presents all those problems and we are gonna need to refactor it. So trying to avoid that same scenario here 🙂

Reply:
I see. So you're suggesting we parse the input environment variable at a higher level into some meaningful struct containing CloudStack credentials and other information, and then pass that struct in when executing cloudmonkey?